### ISSUE TITLE
[PAY 8/19] Fix how validation displays in UI

### ISSUE THREAD

**Issue by Beamanator** [Expensify/App#2934]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
Front-end validation should catch obviously invalid phone numbers, and display the error message `"Please enter a phone number including the country code e.g +447814266907"` instead of a clickable `OptionRow`. Like this:

<img width="50%" alt="Screen Shot 2021-05-14 at 8 52 39 PM" src="https://user-images.githubusercontent.com/3885503/118315862-57079000-b4f6-11eb-930f-99b814f36a54.png">

## Actual Result:
No visual feedback given to user when `OptionRow` with invalid phone number is clicked, only an error is shown in the console:

<img width="90%" alt="Screen Shot 2021-05-14 at 8 49 09 PM" src="https://user-images.githubusercontent.com/3885503/118315481-dba5de80-b4f5-11eb-8323-d7d3e56ec028.png">

## Action Performed:
1. Click search icon
2. Enter an invalid phone number (example: `7777`)
3. Click on the `OptionRow` that shows up:

<img width="50%" alt="Screen Shot 2021-05-14 at 8 47 49 PM" src="https://user-images.githubusercontent.com/3885503/118315331-ad280380-b4f5-11eb-85d6-23f44a7973a3.png">

## Workaround:
No workaround (other than making sure you enter valid phone numbers), but not a deploy-blocking issue.

## Platform:
Where is this issue occurring?

- [x] Web
- [x] iOS
- [x] Android
- [x] Desktop App
- [x] Mobile Web

**Version Number:** 1.0.46.0
**Notes/Photos/Videos:** 
Enhancements that can we worked on concurrently by the assigned engineer of this issue or worked on a week after this closes by someone else https://github.com/Expensify/Expensify.cash/issues/2936 

**Expensify/Expensify Issue URL:** https://github.com/Expensify/Expensify/issues/159892

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @alexpensify (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mallenexpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by Beamanator** [comment]
Unassigning @alexpensify , I don't believe triage is needed :D

**Comment by mallenexpensify** [comment]
Posted to Upwork https://www.upwork.com/jobs/~0175575d61566ad5ee
@Beamanator assigning back to you

**Comment by MelvinBot** [comment]
Triggered auto assignment to @beamanator (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by alev374** [comment]
Hello Client!
It's one day job for me. I can add the validation function and display error message like you want instead of clickable option row. I will catch the onChangeText event with the correct Debounce value, so that I can validate the phone number each input.
Then error message will be shown if any error has occurred. This can be done in each component or customized component.
It will not affect to the redux action, so it will not require to change the busines logic.
That's all.
I am looking forward your message. I can start to work right now.
Best.
Athur.

**Comment by pranshuchittora** [comment]
## Proposal 📄
Front-end validation should catch obviously invalid phone numbers, and display the error message "Please enter a phone number including the country code e.g +447814266907" instead of a clickable OptionRow. Like this:

## Approach 👨🏼‍💻
File of concern : `/src/components/OptionsSelector.js`
1. We need to validate each input from the user.
2. This will increase the number of call to the BE, to fix this we need to `debounce` the user input which will help in reducing the number of calls. 
3. The error state of the input can be handled easily by showing a relevant message to the user as provided in the issue description.

## Testing Strategy🧪
We can mock the call and purposely fail / invalidate the input and then check for the relevant error message text to have been rendered.

## Expected Delivery Time
Approx 1-2 Days.

## Previous Experience 
- https://github.com/Expensify/Expensify.cash/issues/2811

**Comment by Beamanator** [comment]
@alev374 Thanks for [your proposal](https://github.com/Expensify/Expensify.cash/issues/2934#issuecomment-841480975)!

Although I believe I understand where you're going with your proposal, I can't quite tell how you plan to solve this issue. Next time, please clarify exactly how you plan to implement the fix. More specifically, "This can be done in each component or customized component" is not clear enough.

Additionally, "It will not affect to the redux action, so it will not require to change the business logic" makes me wonder if you've looked through our code much, as we currently don't use the `redux` library. Instead, we use the home-grown https://github.com/Expensify/react-native-onyx.

Next time, please try to be more specific with your approach, and maybe think about formatting your thoughts in an easily-consumable format like Pranshu.

**Comment by Beamanator** [comment]
@pranshuchittora Thanks for [your proposal](https://github.com/Expensify/Expensify.cash/issues/2934#issuecomment-841498560) as well!

Thanks for clearly defining the file(s) you plan to change, and the steps you'll take. As you may have noticed, we split the debouncing enhancement into a new issue (found here: https://github.com/Expensify/Expensify.cash/issues/2936) so if you could please adjust your proposal to remove that plan.

Additionally, if an invalid phone number is entered, see in the original post that we want to see the error message `"Please enter a phone number including the country code e.g +447814266907"`, even though the API returns a slightly different error message.

Please let me know when you have adjusted your proposal :)

**Comment by pranshuchittora** [comment]
Thanks for the feedback @Beamanator. I have updated my proposal.
Good Day:)

**Comment by Beamanator** [comment]
Thanks for the quick update @pranshuchittora ! Your proposal looks great, please make sure you also submitted a proposal in Upwork so @mallenexpensify can assign you the job, then please submit a PR when you have time :)

**Comment by pranshuchittora** [comment]
Hi, @Beamanator thanks a lot. I have applied to work on this ticket via UpWork. I will try to push a draft PR by tomorrow 👍🏼

**Comment by mallenexpensify** [comment]
@pranshuchittora , hired you in Upwork 🎉 
@Beamanator will be reviewing your PR once it's ready.

**Comment by Beamanator** [comment]
@pranshuchittora would you mind giving us an update on your progress with this one? I haven't seen a PR yet :)

**Comment by pranshuchittora** [comment]
> @pranshuchittora would you mind giving us an update on your progress with this one? I haven't seen a PR yet :)

Hi, @Beamanator I forgot to raise the PR 😅 . Just raised #3124

**Comment by Beamanator** [comment]
Awesome, thanks!

**Comment by mallenexpensify** [comment]
@Beamanator have you reviewed the PR? Can you and/or @pranshuchittora give an update on status for this issue?

**Comment by pranshuchittora** [comment]
@mallenexpensify waiting for this to be resolved -> https://github.com/Expensify/Expensify.cash/issues/2936#issuecomment-850636244

**Comment by mallenexpensify** [comment]
Thanks @pranshuchittora , I realized after I posted then went to #2936 that that might be the case

**Comment by Beamanator** [comment]
I'll post in the other issue too, but I believe Shawn gave the thumbs-up to push forward with the [latest plan](https://github.com/Expensify/Expensify.cash/issues/2936#issuecomment-850482854)

**Comment by mallenexpensify** [comment]
@Beamanator @pranshuchittora , can I get an update on this? Want to post in the parent issue where we're at.  Thanks

**Comment by Beamanator** [comment]
Waiting on @pranshuchittora to respond to my comments or work on requested changes

**Comment by mallenexpensify** [comment]
Unassigned @pranshuchittora for lack of updates.

**Comment by mallenexpensify** [comment]
Posted new job in Upwork and raised price to $500
https://www.upwork.com/jobs/~018f3274b947f5d7b2

**Comment by rdjuric** [comment]
Hi @Beamanator 

I saw that we were making an API request to the `IsValidPhoneNumber` endpoint in the other PR. Is there any reason we can't use our `Str.isValidPhone` to validate? If we can, I think we could do something like:

1. While the value of our `OptionsSelector` is composed of digits (and +) only, validate it against `Str.isValidPhone`
2. If it's a valid phone number, render the `OptionsList` below the input
3. If it isn't, render an error message

**Comment by aliabbasmalik8** [comment]
I can do this JOB according to [provided plan](https://github.com/Expensify/App/issues/2936#issuecomment-850482854).

Thanks

**Comment by Santhosh-Sellavel** [comment]
## Proposal
Just add a valid in OptionsListUtils.js in below shown method.

<img width="933" alt="Screenshot 2021-07-29 at 3 57 57 AM" src="https://user-images.githubusercontent.com/85645967/127404343-73c2b4f5-6f34-437e-bc68-c0980a20adf0.png">

This is to show the message alone if number is invalid if we don't need show the list, we can do that here
https://github.com/Expensify/App/blob/286b8ea195c445487950e8d8a2cbe86d83bcd997/src/components/OptionsList.js#L215-L228

if `headerMessage` is available we can hide the list.

**Comment by mananjadhav** [comment]
I am sending a proposal for this if it is related to the validation message.  I got a little lost looking at all the discussion points and PR comments. 

**Proposal**
- My approach would be to have a client-side validation and fail early along with the debounce. Preferably Str.isValidPhone as used in the login
- Updates will include: OptionListUtils.
- My other recommendation was if we would want to stick to E164 format for the search at least. It basically matches 2 digit numbers too.
- Add design for the relevant validation messages

**Comment by Beamanator** [comment]
@aliabbasmalik8 
The link you provided is a plan for what will be the outcome, but next time please mention how you plan to accomplish the plan (which files you'll edit, etc)

@Santhosh-Sellavel 
I love the amount of detail you added and even the demo, but your solution is a bit over-engineered. We don't _yet_ need a new type of phone number validation (E.123). Keep up the fantastic proposals though 💯 

@mananjadhav 
Thanks for your proposal, but in order to be selected for the job I'd like to see more specifics about what you're proposing (ex: what exactly will you change in OptionListUtils?). Please include proposals for only this issue as well (note: debouncing is a requirement for a different issue).

**Comment by Beamanator** [comment]
@rdjuric 
Good question, eventually we will want to call `IsValidPhoneNumber` because it connects to a `Twilio` api that validates a phone number exists, which is better validation than just a regular expression. However, we won't need to handle that to fix this PR.

Your solution is very simple and straightforward, I like it 👍 Please submit a PR when you have a moment and @mallenexpensify will assign you the job in Upwork!

**Comment by Beamanator** [comment]
Contrib linked PR, changing to Weekly

**Comment by rdjuric** [comment]
@Beamanator Pullerbear didn't assign you to review the PR and I think I might need some help/inputs from you there.

Can you take a look at it when you have time?

**Comment by Beamanator** [comment]
PR was merged but @rdjuric agreed there's one final fix needed, he said he'll open up a PR soon (see [this comment](https://github.com/Expensify/App/pull/4354#issuecomment-893290019))

**Comment by rdjuric** [comment]
Follow-up PR is ready for review 😃

**Comment by Beamanator** [comment]
Aaand merged 💪

**Comment by mallenexpensify** [comment]
Will pay onn 8/19 if no regressions

**Comment by Santhosh-Sellavel** [comment]
@rdjuric or @Beamanator  I'm still able to reproduce this issue.

https://user-images.githubusercontent.com/85645967/129308620-cc3752d2-cc0e-403b-a49c-83e92e38554d.mov

`DIGITS_AND_PLUS: /^\+?[0-9]*$/,`
In above expression still `+` is still optional.


`E164_REGEX: /^\+?[1-9]\d{1,14}$/` which is used by `Str.isValidPhone `method,` + `is also optional here

Consider below expression  `Str.isValidPhone(searchValue)` will always be true by using `!` over it makes the below expression always false, so no error message will be returned.

<img width="919" alt="Screenshot 2021-08-13 at 10 45 54 AM" src="https://user-images.githubusercontent.com/85645967/129308733-70e39493-0a9d-4af5-a51b-c418808b2464.png">

**Comment by Santhosh-Sellavel** [comment]
@Beamanator 
Regex should be updated for DIGITS_AND_PLUS as `/^\+(?:[0-9]*)$/` 

Validation expression should be like below,
`if(searchValue && Str.isValidPhone(searchValue) && !CONST.REGEX. DIGITS_AND_PLUS.test(searchValue))` 

Similar to what I proposed earlier here https://github.com/Expensify/App/issues/2934#issuecomment-888664379, where I suggested `ETU.123` standard regex, instead we could use this `/^\+(?:[0-9]*)$/` for just phone number should have just one `+`

**Comment by Beamanator** [comment]
@Santhosh-Sellavel did you try clicking on the "new chat" option (+911223233) at the end of your video? If so, you should have seen the "I couldn't validate the phone number, please try again" error -> that's the main fix for this issue. Improving the regex is part of a different issue that I plan to write the requirements for early next week

**Comment by Santhosh-Sellavel** [comment]
Yes, I'm seeing that error on click,
<img width="716" alt="Screenshot 2021-08-13 at 5 03 21 PM" src="https://user-images.githubusercontent.com/85645967/129351322-88567fbd-ec76-4475-8e6d-e691c16dbb7f.png">

Sorry, I am unaware of the requirement change here! 
There were two PR's that caused me the confusion,  where one looks like solving validation message, the initial requirement.

Thanks for the clarity mate! @Beamanator

**Comment by Beamanator** [comment]
Yeah sorry for the confusion @Santhosh-Sellavel , technically both requirements were there from the beginning but

**Comment by Beamanator** [comment]
Wrong button 😅

**Comment by rdjuric** [comment]
Hi @mallenexpensify! Is there any job in Upwork for this issue? I wasn't hired yet and can't find one

**Comment by mallenexpensify** [comment]
@rdjuric lotta action in here, let's assume I messed something up. 
1. The most recent job post is here https://www.upwork.com/jobs/~018f3274b947f5d7b2
2. Is this you? If so, confirm and I'll hire 
![image](https://user-images.githubusercontent.com/22508304/129958727-cbadd13d-34cf-44e7-898a-8e5766ca67b7.png)

**Comment by rdjuric** [comment]
@mallenexpensify That's me! Thanks.

**Comment by mallenexpensify** [comment]
Hired @rdjuric in Upwork!

**Comment by mallenexpensify** [comment]
Paid in Upwork!

**Comment by Beamanator** [comment]
Great! Looks like it's time to close this out :D
